sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java

```diff
@@ -945,4 +951,132 @@ public static CustomHttpErrors createBigQueryClientCustomErrors() {
             + "https://cloud.google.com/service-infrastructure/docs/rate-limiting#configure");
     return builder.build();
   }
+
+  private static class BoundedExecutorService implements ExecutorService {
+    private ExecutorService executor;
+    private Semaphore semaphore;
+
+    BoundedExecutorService(ExecutorService executor, int parallelism) {
+      this.executor = executor;
+      this.semaphore = new Semaphore(parallelism);
+    }
+
+    @Override
+    public void shutdown() {
+      executor.shutdown();
+    }
+
+    @Override
+    public List<Runnable> shutdownNow() {
```

Comment:
A bit weird that what will get returned here are the wrapped SemaphoreRunnable objects. Technically this violates the contract of Executor (you should get back Executors that you put in)

Reply:
Yes, you're right. Ideally we should return runnable objects which have the same behavior as we put in. However it seems impossible because the delegated `shutdownNow()` doesn't only return `SemaphoreRunnable` but also `RunnableFuture` that there's no way to access to its original `SemaphoreCallable` object. I think we can allow the semaphore object to have the maximum number of permits before returning from `shutdownNow()` as less-ideal alternative.